torch/distributed/_tools/memory_tracker.py

```diff
@@ -0,0 +1,246 @@
+        self.markers: Dict[str, int] = defaultdict(int)
+        self.cur_module: str = ""
+        self.op_id: int = 0
+
+    @no_type_check
+    def start_monitor(self, root_module: nn.Module) -> None:
+        """
+        Register module hooks and entering MemoryProfileDispatchMode, so that
+        module level and operator level memory stats can be tracked during module
+        runtime.
+        """
+        self._clear_state()
+        root_module.__setattr__("is_root", True)
+        for name, m in root_module.named_modules():
+            if m is not root_module:
+                m.__setattr__("is_root", False)
+            # fused_proxy_group does not support hooks
+            if ".fused_proxy_grouped_embedding_bag" in name:
+                continue
+            h1 = m.register_forward_pre_hook(self._create_pre_forward_hook(name))
+            h2 = m.register_forward_hook(self._create_post_forward_hook(name))
+            h3 = m.register_backward_hook(self._create_backward_hook(name))
+            self._hooks.extend([h1, h2, h3])
+        torch.cuda.empty_cache()
+        assert getattr(self, "profile_mode", None) is None
+        self.profile_mode = MemoryProfileDispatchMode(self)
+        self.profile_mode.__enter__()
+
+    @no_type_check
+    def stop(self) -> None:
+        """
+        Remove module hooks and exit MemoryProfileDispatchMode to stop
+        tracking memory stats at module level and operator level.
+        """
+        for h in self._hooks:
+            h.remove()
+        self._hooks.clear()
+        assert getattr(self, "profile_mode", None) is not None
+        self.profile_mode.__exit__(None, None, None)
+        self.profile_mode = None
+
+    @no_type_check
+    def summary(self, top: int = 20) -> None:
+        """
+        Print out the top 20 operators that generate the most memories.
+        """
+        op_diff: Dict[str, float] = defaultdict(float)
+        op_name, previous_allocated_memory = self.memories_allocated[0]
+        for i in range(1, self.op_id):
+            op_name, current_allocated_memory = self.memories_allocated[i]
+            op_diff[op_name] = current_allocated_memory - previous_allocated_memory
+            previous_allocated_memory = current_allocated_memory
+
+        print("------------------------------------------------")
+        print(f"Top {top} ops that generates memory are:")
+        for k, v in sorted(op_diff.items(), key=lambda item: item[1], reverse=True)[
+            :top
+        ]:
+            print(f"{k}: {v}MB")
+        print("------------------------------------------------")
+
+    @no_type_check
+    def show_traces(self) -> None:
+        """
+        Show the traces of memory_allocated, memory_active and memory_reserved at
+        operator level and the marker 'fw_bw_boundary' at the boundary of forward pass
+        and backward pass.
+        """
+        import matplotlib.pyplot as plt
```

Comment:
Should we have some better error handling around this import?

Reply:
I think it is OK to crash if the import failed?